### ISSUE TITLE
500 errors when attempting to edit/move a page where Page.specific_page returns None

### ISSUE THREAD

**Issue by sdonk** [wagtail/wagtail#5209]
### Issue Summary

Instances of deleted page types cannot be edited or deleted in the admin UI.

### Steps to Reproduce

1. Add a new page type in `models.py` and run the relative migration
2. Add a page instance of the above type in the wagtail admin UI
3. Delete the page type from `models.py` and run the relative migration
4. The page instance appears in the navigation tree, with an empty type, but it cannot be deleted in the UI and when trying to edit it a 500 error is returned

The page can be deleted in the shell. I was expecting the delete button to be available in the UI and the edit link being disabled.

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: no

### Technical details

* Python version: 3.6.8
* Django version: 1.11.20
* Wagtail version: 2.3

**Comment by ababic** [comment]
Thanks for reporting this @sdonk. Just investigating this now :)

**Comment by ababic** [comment]
In the bakery demo, using the latest version of Wagtail, following the above steps does not result in the delete option not being available, but I can confirm there are 500 errors when you attempt some things (see further down for more info).

Here's how the explorer looks after removing the custom page model:

![Screen Shot 2019-04-09 at 5 05 13 pm](https://user-images.githubusercontent.com/7779588/55816557-67183e80-5aea-11e9-8767-a916597afb3f.png)

## When trying to edit the page:

The edit view seems to work fine until you try to save a draft. Upon reloading the page, Wagtail fails with the following traceback:

```
File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.5/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.5/site-packages/django/core/handlers/base.py" in _get_response
  126.                 response = self.process_exception_by_middleware(e, request)

File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.5/site-packages/django/core/handlers/base.py" in _get_response
  124.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.5/site-packages/django/views/decorators/cache.py" in _wrapped_view_func
  44.         response = view_func(request, *args, **kwargs)

File "/vagrant/wagtail/wagtail/admin/urls/__init__.py" in wrapper
  102.             return view_func(request, *args, **kwargs)

File "/vagrant/wagtail/wagtail/admin/decorators.py" in decorated_view
  34.             return view_func(request, *args, **kwargs)

File "/vagrant/wagtail/wagtail/admin/views/pages.py" in edit
  320.     page = real_page_record.get_latest_revision_as_page()

File "/vagrant/wagtail/wagtail/core/models.py" in get_latest_revision_as_page
  687.             return latest_revision.as_page_object()

File "/vagrant/wagtail/wagtail/core/models.py" in as_page_object
  1466.         obj = self.page.specific_class.from_json(self.content_json)

Exception Type: AttributeError at /admin/pages/81/edit/
Exception Value: 'NoneType' object has no attribute 'from_json'
```

## When trying to move the page:

After using the view to choose a target page as the new parent, Wagtail fails with the following traceback:

```
File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.5/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.5/site-packages/django/core/handlers/base.py" in _get_response
  126.                 response = self.process_exception_by_middleware(e, request)

File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.5/site-packages/django/core/handlers/base.py" in _get_response
  124.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/home/vagrant/.virtualenvs/bakerydemo/lib/python3.5/site-packages/django/views/decorators/cache.py" in _wrapped_view_func
  44.         response = view_func(request, *args, **kwargs)

File "/vagrant/wagtail/wagtail/admin/urls/__init__.py" in wrapper
  102.             return view_func(request, *args, **kwargs)

File "/vagrant/wagtail/wagtail/admin/decorators.py" in decorated_view
  34.             return view_func(request, *args, **kwargs)

File "/vagrant/wagtail/wagtail/admin/views/pages.py" in move_choose_destination
  734.         target.can_choose = page_perms.can_move_to(target)

File "/vagrant/wagtail/wagtail/core/models.py" in can_move_to
  1836.         if not self.page.specific.can_move_to(destination):

File "/vagrant/wagtail/wagtail/core/models.py" in can_move_to
  1003.         return self.can_exist_under(parent)

File "/vagrant/wagtail/wagtail/core/models.py" in can_exist_under
  980.         return cls in parent.specific_class.allowed_subpage_models()

Exception Type: AttributeError at /admin/pages/82/move/3/
Exception Value: 'NoneType' object has no attribute 'allowed_subpage_models'

**Comment by sdonk** [comment]
In my case the page had a live version and a draft version when the relative model was deleted.

This is what I get when I try to hover it (it's grey because of that)
![Screenshot 2019-04-09 at 17 21 04](https://user-images.githubusercontent.com/2109971/55817333-346f4580-5aec-11e9-8b4a-ceb552c47cf1.jpg)

**Comment by ababic** [comment]
Thanks @sdonk. Yes, if draft version already existed, I imagine you'd get stack trace very similar to the first one here when attempting to edit. I'll look into this some more to see what the best fix here would be.

**Comment by sdonk** [comment]
The stacktrace I get when I try to edit the page is pretty much the same you got

**Comment by ababic** [comment]
I would propose a three-fold fix here:

### 1. Make it possible to save drafts and publish changes to a vanilla `Page` object

Although the page type has been deleted, being able to edit the title or other base fields for the page is arguably still useful.

### 2. Have the edit view gracefully handle cases where a draft is for a different type of page

I think abandoning the draft and loading the saved page state instead would work, with a clear warning that says something like...

> Wagtail cannot load previous draft content for this page because the original page type has been removed. The latest published content is shown instead.

(When the page type has been removed)

OR

> Wagtail cannot load previous draft content for this page because the draft is no longer compatible.  This usually happens as a result of pages being modified outside of Wagtail. The latest published content is shown instead.

(When the page type has changed somehow)

### 3. Make it possible to move a vanilla `Page` object

Arguably, we could disable this somehow instead. But, I'm guessing that fixing the above error would be more helpful (and less work)

**Comment by gasman** [comment]
Addressed in #5240. As per the discussion on that PR, a missing page model _is_ an error condition, and one that should be communicated to the developer rather than the end user, and so it's appropriate that Wagtail returns a 500 error for actions that can't be completed with a missing page model. However, as of #5240 there is now a more descriptive error message for this case.
